nimare/meta/utils.py

```diff
@@ -16,6 +16,125 @@
 LGR = logging.getLogger(__name__)
 
 
+def _calculate_cluster_measures(arr3d, threshold, conn, tail="upper"):
+    """Calculate maximum cluster mass and size for an array.
+
+    This method assesses both positive and negative clusters.
+
+    Parameters
+    ----------
+    arr3d : :obj:`numpy.ndarray`
+        Unthresholded 3D summary-statistic matrix. This matrix will end up changed in place.
+    threshold : :obj:`float`
+        Uncorrected summary-statistic thresholded for defining clusters.
+    conn : :obj:`numpy.ndarray` of shape (3, 3, 3)
+        Connectivity matrix for defining clusters.
+
+    Returns
+    -------
+    max_size, max_mass : :obj:`float`
+        Maximum cluster size and mass from the matrix.
+    """
+    if tail == "upper":
+        arr3d[arr3d <= threshold] = 0
+    else:
+        arr3d[np.abs(arr3d) <= threshold] = 0
+
+    labeled_arr3d, _ = ndimage.measurements.label(arr3d > 0, conn)
+
+    if tail == "two":
+        # Label positive and negative clusters separately
+        n_positive_clusters = np.max(labeled_arr3d)
+        temp_labeled_arr3d, _ = ndimage.measurements.label(arr3d < 0, conn)
+        temp_labeled_arr3d[temp_labeled_arr3d > 0] += n_positive_clusters
+        labeled_arr3d = labeled_arr3d + temp_labeled_arr3d
+        del temp_labeled_arr3d
+
+    clust_vals, clust_sizes = np.unique(labeled_arr3d, return_counts=True)
+    assert clust_vals[0] == 0
+
+    # Cluster mass-based inference
+    max_mass = 0
+    for unique_val in clust_vals[1:]:
+        ss_vals = np.abs(arr3d[labeled_arr3d == unique_val]) - threshold
+        max_mass = np.maximum(max_mass, np.sum(ss_vals))
+
+    # Cluster size-based inference
+    clust_sizes = clust_sizes[1:]  # First cluster is zeros in matrix
+    if clust_sizes.size:
+        max_size = np.max(clust_sizes)
+    else:
+        max_size = 0
+
+    return max_size, max_mass
+
+
+def _get_last_bin(arr1d):
+    """Index the last location in a 1D array with a non-zero value."""
+    if np.any(arr1d):
+        last_bin = np.where(arr1d)[0][-1]
+
+    else:
+        last_bin = 0
+
+    return last_bin
+
+
+def calculate_tfce(z_map, E=0.5, H=2, dh=0.1):
```

Comment:
@nicholst I was hoping you could take a look at this function to see if it calculates TFCE correctly.